File: ios/Plugin/FirebaseAuthenticationHelper.swift
import Foundation
import Capacitor
import FirebaseCore
import FirebaseAuth

public class FirebaseAuthenticationHelper {
    public static func createSignInResult(credential: AuthCredential, user: User?, nonce: String?) -> JSObject {
        let userResult = self.createUserResultFromFirebaseUser(user)
        let credentialResult = self.createCredentialResultFromAuthCredential(credential, nonce: nonce)
        var result = JSObject()
        result["user"] = userResult
        result["credential"] = credentialResult
        return result
    }

    public static func createUserResultFromFirebaseUser(_ user: User?) -> JSObject? {
        if user == nil {
            return nil
        }
        var result = JSObject()
        result["displayName"] = user?.displayName
        result["email"] = user?.email
        result["emailVerified"] = user?.isEmailVerified
        result["isAnonymous"] = user?.isAnonymous
        result["phoneNumber"] = user?.phoneNumber
        result["photoUrl"] = user?.photoURL?.absoluteString
        result["providerId"] = user?.providerID
        result["tenantId"] = user?.tenantID
        result["uid"] = user?.uid
        return result
    }

    public static func createCredentialResultFromAuthCredential(_ credential: AuthCredential?, nonce: String?) -> JSObject? {
        guard let credential = credential else {
            return nil
        }
        var result = JSObject()
        result["providerId"] = credential.provider
        
        if let oAuthCredential = credential as? OAuthCredential {
            // If it's OAuthCredential, we can get its values directly
            
            let accessToken = oAuthCredential.accessToken
            if accessToken != nil {
                result["accessToken"] = accessToken
            }
            let idToken = oAuthCredential.idToken
            if idToken != nil {
                result["idToken"] = idToken
            }
            let secret = oAuthCredential.secret
            if secret != nil {
                result["secret"] = secret
            }
        } else {
            // However, a AuthCredential doesn't expose idToken/accessToken that
            // we need. So we use KVC to try and get them.
        
            if let idToken = credential.value(forKey: "_IDToken") as? String {
                result["idToken"] = idToken
            }
            if let accessToken = credential.value(forKey: "_accessToken") as? String {
                result["accessToken"] = accessToken
            }
        }
        
        if let nonce = nonce {
            result["nonce"] = nonce
        }
        return result
    }
}

File: android/src/main/java/dev/robingenz/capacitorjs/plugins/firebase/auth/FirebaseAuthenticationHelper.java
package dev.robingenz.capacitorjs.plugins.firebase.auth;

import com.getcapacitor.JSObject;
import com.google.firebase.auth.AuthCredential;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.auth.GetTokenResult;
import com.google.firebase.auth.GoogleAuthCredential;
import com.google.firebase.auth.OAuthCredential;
import com.google.firebase.auth.PhoneAuthCredential;
import dev.robingenz.capacitorjs.plugins.firebase.auth.handlers.GoogleAuthProviderHandler;
import dev.robingenz.capacitorjs.plugins.firebase.auth.handlers.OAuthProviderHandler;

public class FirebaseAuthenticationHelper {

    public static JSObject createSignInResult(FirebaseUser user, AuthCredential credential, String idToken) {
        JSObject userResult = FirebaseAuthenticationHelper.createUserResult(user);
        JSObject credentialResult = FirebaseAuthenticationHelper.createCredentialResult(credential, idToken);
        JSObject result = new JSObject();
        result.put("user", userResult);
        result.put("credential", credentialResult);
        return result;
    }

    public static JSObject createUserResult(FirebaseUser user) {
        if (user == null) {
            return null;
        }
        JSObject result = new JSObject();
        result.put("displayName", user.getDisplayName());
        result.put("email", user.getEmail());
        result.put("emailVerified", user.isEmailVerified());
        result.put("isAnonymous", user.isAnonymous());
        result.put("phoneNumber", user.getPhoneNumber());
        result.put("photoUrl", user.getPhotoUrl());
        result.put("providerId", user.getProviderId());
        result.put("tenantId", user.getTenantId());
        result.put("uid", user.getUid());
        return result;
    }

    public static JSObject createCredentialResult(AuthCredential credential, String idToken) {
        if (credential == null && idToken == null) {
            return null;
        }
        JSObject result = new JSObject();
        if (credential != null) {
            result.put("providerId", credential.getProvider());
        }
        if (credential instanceof OAuthCredential) {
            String oAuthAccessToken = ((OAuthCredential) credential).getAccessToken();
            if (oAuthAccessToken != null) {
                result.put("accessToken", oAuthAccessToken);
            }
            String oAuthIdToken = ((OAuthCredential) credential).getIdToken();
            if (oAuthIdToken != null) {
                result.put("idToken", oAuthIdToken);
            }
            String oAuthSecret = ((OAuthCredential) credential).getSecret();
            if (oAuthSecret != null) {
                result.put("secret", oAuthSecret);
            }
        } else if (idToken != null) {
            result.put("idToken", idToken);
        }
        return result;
    }
}
